ImperatorToCK3/CK3/World.cs

```diff
@@ -314,40 +363,95 @@ Imperator.World impWorld
+					continue;
+				}
+				var ck3CapitalBaronyProvince = Provinces[title.CapitalBaronyProvince];
+				var impProvince = ck3CapitalBaronyProvince.ImperatorProvince;
+
+				if (impProvince is not null) {
+					var impCountry = impProvince.OwnerCountry.Value;
+
+					if (impCountry is not null && impCountry.CountryType != CountryType.rebels) {
+						var ck3Country = impCountry.CK3Title;
+						if (ck3Country is null) {
+							Logger.Warn($"{impCountry.Name} has no CK3 title!"); // should not happen
+							continue;
+						}
+						var ck3CapitalCounty = ck3Country.CapitalCounty;
+						var impMonarch = impCountry.Monarch;
+						var matchingGovernorships = new List<Governorship>(governorships.Where(g =>
+							g.CountryID == impCountry.ID &&
+							g.RegionName == imperatorRegionMapper.GetParentRegionName(impProvince.ID)
+						));
+
+						if (!ck3CapitalCounty.HasValue || ck3CapitalCounty.Value.Value is null) {
+							if (impMonarch is not null) {
+								GiveCountyToMonarch(title, ck3Country, (ulong)impMonarch, ck3BookmarkDate);
+							} else {
+								Logger.Warn($"Imperator ruler doesn't exist for {impCountry.Name} owning {title.Name}!");
+							}
+							continue;
+						}
+						// if title belongs to country ruler's capital's de jure duchy, make it directly held by the ruler
+						var countryCapitalDuchy = ck3CapitalCounty.Value.Value.DeJureLiege;
+						var titleLiegeDuchy = title.DeJureLiege;
+						if (countryCapitalDuchy is not null && titleLiegeDuchy is not null && countryCapitalDuchy.Name == titleLiegeDuchy.Name) {
+							if (impMonarch is not null) {
+								GiveCountyToMonarch(title, ck3Country, (ulong)impMonarch, ck3BookmarkDate);
+							}
+						} else if (matchingGovernorships.Count > 0) {
+							// give county to governor
+							var governorship = matchingGovernorships[0];
+							var ck3GovernorshipName = tagTitleMapper.GetTitleForGovernorship(governorship.RegionName, impCountry.Tag, ck3Country.Name);
+							if (ck3GovernorshipName is null) {
+								Logger.Warn(nameof(ck3GovernorshipName) + $" is null for {ck3Country.Name} {governorship.RegionName}!");
+								continue;
 							}
-						} else { // county is probably outside of Imperator map
-							var holderId = title.GetHolderId(ck3BookmarkDate);
-							if (!string.IsNullOrEmpty(holderId) && holderId != "0") {
-								countyHoldersCache.Add(holderId);
+							var ck3Governorship = LandedTitles[ck3GovernorshipName];
+							if (Characters.TryGetValue(ck3Governorship.GetHolderId(ck3BookmarkDate), out var governor)) {
+								title.ClearHolderHistory();
+								title.SetHolderId(governor.ID, ck3Governorship.GetDateOfLastHolderChange());
 							}
+							title.DeFactoLiege = null;
+							countyHoldersCache.Add(title.GetHolderId(ck3BookmarkDate));
+						} else if (impMonarch is not null) {
+							GiveCountyToMonarch(title, ck3Country, (ulong)impMonarch, ck3BookmarkDate);
```

Comment:
This is long enough you should consider extracting some functions

Reply:
Extracted `GiverCountryToGovernor`.